### ISSUE TITLE
Possible to emulate request to version.json + possible steps for full offline support?

### ISSUE THREAD

**Issue by shogunator** [Melledy/Grasscutter#664]
Related: #326 

I'm trying to get Grasscutter + game client to work completely offline.

So far the steps I've taken are:

- Run a local DNS server (https://maradns.samiam.org/)
- Configure the ipv4 names similar to proxy.py (https://github.com/Grasscutters/Grasscutter/blob/a185cd167ea38a61fcc0aac1d286377cb89e335c/proxy.py)
- Add a virtual network adapter in Windows, set the DNS server to 127.0.0.1 so MaraDNS is able to resolve the requests

At this point the client opens and does not give me a 4201 error anymore even without an internet connection, however it is now stuck at "Checking for updates" then times out.

Checking the log files in AppData/LocalLow for the game show a request to:

https://webstatic-sea.hoyoverse.com/admin/mi18n/plat_oversea/m2020030410/m2020030410-version.json

Response is:

`{"version":65}`

The error in the log file is just `error=Generic/unknown HTTP error`.

---

Checking if Grasscutter has that route defined in a browser + DevTools, it simply gives the 404 not found page instead of the JSON.

---

Edit: it seems that the server has a version hardcoded in the code. Is it possible to make this configurable?

Relevant line: https://github.com/Grasscutters/Grasscutter/commit/582d91227cf14d729954e8c78a919e265913cdef

I added a route for it on my local repository & build, and got past the initial error 4201 (unable to connect)

---

Edit 2: there's also a HTTP (not HTTPS) request when the game tries to download an update.

```
{"remoteName": "blocks/00/22551915.blk", "md5": "0cf3d6b599d443c7cfa4a3e6189a7757", "fileSize": 3184}
{"remoteName": "blocks/00/25060239.blk", "md5": "f3cf18d697e3380b2f833ccf2c7d4194", "fileSize": 3328}
{"remoteName": "blocks/00/29342328.blk", "md5": "bc652abe9338f1b94f112ad5f8a2e16d", "fileSize": 11245}
{"remoteName": "blocks/00/32070509.blk", "md5": "4bf2a81d7565301269d1c90e14393045", "fileSize": 698}
{"remoteName": "blocks/00/33067900.blk", "md5": "0b4d781c1633537d2fa4b4f0f567d2ca", "fileSize": 454}
```

http://autopatchhk.yuanshen.com/client_design_data/2.6_live/output_6731353_adfeb8be71/client_silence/General/AssetBundles/data_versions

**Comment by AZthemute** [comment]
Have you tried analysing what happens when the client works normally with the official servers and no proxy nonsense?

**Comment by shogunator** [comment]
I set these up for testing:

![image](https://user-images.githubusercontent.com/105140466/167285549-9609bd93-2ef8-4da3-bc76-aee13913182f.png)

![image](https://user-images.githubusercontent.com/105140466/167285556-841dde2a-5ea9-499a-96b0-d843be6f4e73.png)

~~The request is going through Teletrik (can see the Tunnel on the initial connection & the follow-up request to the endpoint for `client_design_data`)~~

Opening the page on Edge works successfully.

The error message the game client sends back to the `/log` endpoint on the Dispatch Server:

```JavaScript
{
    "userName": "Test",
    "time": "2022-05-08 14:06:25.5464",
    "frame": "",
    "stackTrace": "MoleMole.SuperDebug:LogToServerInternal(Boolean, String, LogType, Boolean, Int32, Int32, ErrorLevel, ErrorCategory, String, String)\nMoleMole.SuperDebug:LogToServer(LogType, String, Boolean, Int32, ErrorLevel, ErrorCategory, String, Boolean, Int32, String)\nMoleMole.SuperDebug:VeryImportantError(String, ErrorLevel, ErrorCategory, String, Boolean, Int32, Int32, String)\nEKAOONLAJML:EKNGAGLAGDO()\nGILBIFNGGCL:Tick()\nJMCCFDKDILB:EKNGAGLAGDO()\nMoleMole.GameManager:Update()\n",
    "logStr": "FileManageSystem=> download file failed, url = https://autopatchhk.yuanshen.com/client_design_data/2.6_live/output_6731353_adfeb8be71/client_silence/General/AssetBundles/data_versions, localPath = C:/Program Files/Genshin Impact/Genshin Impact game/GenshinImpact_Data/Persistent\\silence_data_versions_remote, errMsg = -9203-Error getting response stream (Write: The authentication or decryption has failed.): SendFailureremote ip: 127.0.0.1.host: autopatchhk.yuanshen.com; , useTime = 15s",
    "logType": "Error",
    // ...
}
```

---

I *can* log in while I'm connected to the internet, but I'm trying to get this working fully offline at the moment.

---

Edit: set the headers to 'application/octet-stream' similar to the original response from autopatchhk, and still not working. Perhaps there is some certificate pinning?

---

Edit 2: ~~the request doesn't seem to be MITMed properly by Fiddler when the game makes the request.~~

~~When I make the request by following the URL on Edge it is MITMed and shows up on Fiddler.~~

See my latest comment, this is due to a misconfiguration on Grasscutter/server-side where it sends back information to the client to connect to 127.0.0.1 instead of the actual proxied address.

**Comment by leonier** [comment]
I think later we have to block any major upgrade request from mhy servers in case that mhy strengthened encryption in 2.7.

**Comment by krrr** [comment]
you can use my mitmproxy script, it dose not require local DNS
```python
"""yuanshen mitm. need improve."""
from mitmproxy import http
from mitmproxy import ctx
from mitmproxy.proxy import layer, layers


def load(loader):
    ctx.options.web_open_browser = False
    # We change the connection strategy to lazy so that next_layer happens before we actually connect upstream.
    ctx.options.connection_strategy = "lazy"
    ctx.options.upstream_cert = False
    ctx.options.ssl_insecure = True

    
def next_layer(nextlayer: layer.NextLayer):
    ctx.log(
        f"{nextlayer.context=}\n"
        f"{nextlayer.data_client()[:70]=}\n"
    )
    sni = nextlayer.context.client.sni
    if nextlayer.context.client.tls and sni and (sni.endswith("yuanshen.com") or sni.endswith("mihoyo.com") or sni.endswith("hoyoverse.com")):
        ctx.log('sni:' + sni)
        nextlayer.context.server.address = ("127.0.0.1", 443)


def request(flow: http.HTTPFlow) -> None:
    # pretty_host takes the "Host" header of the request into account
    if flow.request.pretty_url.startswith('http://log-upload-os.mihoyo.com'):
        flow.response = http.Response.make(
            404,  # (optional) status code
            b"404 not found",  # (optional) content
            {"Content-Type": "text/html"}  # (optional) headers
        )
        return
```

**Comment by shogunator** [comment]
I can confirm the latest `development` branch can run fully offline (tested commit e0b84e275).

One thing I had to change on my side (to get past the 4206) was as follows https://github.com/Grasscutters/Grasscutter/issues/120#issuecomment-1107412320

> Usually this is because of query_region_list returned an incorrect address. Try to change the dispatchpublicip to dispatchosglobal.yuanshen.com or dispatchcnglobal.yuanshen.com, the proxy daemon will automatically process this address to the correct host.

On `config.json`, change `server.dispatch.accessAddress` and `server.game.accessAddress` to `dispatchosglobal.yuanshen.com`.

The reason I was getting 4206 errors and the traffic not going to MITM proxy was I set those values to 127.0.0.1. 

- Game hits /query_region_list, gets the dispatch address of 127.0.0.1
- 127.0.0.1 is not MITMed by the proxy. Grasscutter's serves HTTPS certificate using keystore.p12 (which is invalid for 127.0.0.1 anyways, since the subject or SAN is not 127.0.0.1)
- Once the values in config.json were changed, the game now does the request to dispatchosglobal.yuanshen.com, which is MITMed and a "valid" certificate presented to the game by MITMProxy/Teletrik)

@krrr's script also works great for the proxy. I did comment out the `ctx.options.web_open_browser = False` part since it caused an error on my mitmdump (config option does not exist, maybe I'm using an older version of mitmproxy?)
